keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -425,9 +436,10 @@ type IngressBackend struct {
 }
```

Comment:
> The API server will automatically populate the Resource field below.
> ...
> If both ServiceName and Resource are specified, then they must reference the same Service object.

this sounds like cross-field defaulting, which got us in trouble with workload label selectors. consider the following scenario:
1. user creates ingress not knowing about the new fields, populating service name with "foo"
2. the apiserver populates resource with `type: Service, name: foo, apiGroup: ""`
3. the user modifies the ingress to change service name to "bar"
4. apiserver fails validation because service name and resource.name no longer match

Reply:
Agree.  Can we call this a one-of without a discriminator (we have high confidence it will only be 2 options)?  Do we need to structure it as below?

```
  BackendType string
  Service ServicBackend // {name string, port int or string}
  Resource v1.TypedLocalObjectReference
```

Or can we leave it like this get away with it?

@apelisse

Reply:
I'd also like to consider not using IntOrString and instead doing something like:

```
Port BackendPort
```
```
type BackendPort {
    Number int
    Name string
}
```

Do we need a discriminator if we know it is A or B and there will never be a C ?

Reply:
Probably not, as all clients know they have to reset the other value. 

As a side note, `Name` is usually used as a key for associative lists though, so that's slightly confusing outside the context.

Reply:
@thockin I would even like to have just an `int32`, but I think this kind of change is bigger than `beta -> GA`

Reply:
I've changed the proposal to have a one-of union between ServiceBackend, Resource. If someone wants to reference a Service using the Resource, that should still be ok, just more verbose for them.

@thockin -- is IntOrString considered something we wish to move away from?

Reply:
IntOrString is generally seen as a mistake at this point.

If I interpret @apelisse correctly, he wants a discriminator field. Is that right?

Reply:
I think they can get away without the discriminator since there are no other non-union members in that structure and it's not embedded anywhere.

My "Probably not" was an answer to:
> Do we need a discriminator if we know it is A or B and there will never be a C ?